### ISSUE TITLE
Set OP_NO_COMPRESSION & OP_CIPHER_SERVER_PREFERENCE on TLS contexts

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6801]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)| @hynek reported|
|-|-|
|Trac ID|trac#6801|
|Type|enhancement|
|Created|2013-10-25 12:37:51Z|
|Branch|https://github.com/twisted/twisted/tree/harden-tls-options-6801|

There is a broad consensus that there are two options that should always be set:

 * OP_NO_COMPRESSION (CRIME attacks, seems like most OpenSSLs set that as default nowadays but still)
 * OP_CIPHER_SERVER_PREFERENCE (if we push for secure ciphers, we should make sure they get used too)

Both are missing in pyOpenSSL [at the moment](https://bugs.launchpad.net/pyopenssl/+bug/1244201) but we can easily wing it for now.

Confusingly, OP_NO_COMPRESSION is documented in my self-built pyOpenSSL docs ( http://glui.me/?i=3b37w0j9wemcrsl/2013-10-25_at_14.30.png – maybe I’ve built them from a bzr checkout? been a while) and my OpenSSL should have support for it but I can’t find it neither in the official docs at http://pythonhosted.org/pyOpenSSL/openssl-ssl.html nor in the actual library. In any case we’d have to take into account that the underlying (Py)OpenSSL may not support either.

Implementation is trivial, I’d just wait for [#6772](https://github.com/twisted/twisted/issues/6772) to  land in trunk because it’s laying groundwork for setting options.

<details><summary>Searchable metadata</summary>

```
trac-id__6801 6801
type__enhancement enhancement
reporter__hynek hynek
priority__normal normal
milestone__ 
branch__branches_harden_tls_options_6801 branches/harden-tls-options-6801
branch_author__hynek hynek
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1382704671000000 1382704671000000
changetime__1389896799000000 1389896799000000
version__None None
owner__hynek hynek
cc__oberstet
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="oberstet's avatar" src="https://avatars.githubusercontent.com/u/233340?s=50" width="50" height="50">](https://github.com/oberstet)<a name="note_1"></a>|@oberstet commented|
|-|-|

Probably these 2 should be set also:

```
SSL.OP_SINGLE_ECDH_USE = 0x00080000L
SSL.OP_SINGLE_DH_USE = 0x00100000L
```

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_2"></a>|@hynek commented|
|-|-|

(In [41219]) Branching to harden-tls-options-6801.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_3"></a>|@hynek commented|
|-|-|

I’ve done the following:

 1. Consolidated options handling by baking them into `self._options` on `__init__` to avoid repeated calls to set_options in `_makeContext`.
 1. Set `OP_NO_COMPRESSION` and `OP_CIPHER_SERVER_PREFERENCE` unconditionally.
 1. Bonus: Added tests that `OP_SINGLE_DH_USE` is set if `enableSingleUseKeys` is active and also added `OP_SINGLE_ECDH_USE`in the same case so we don't forget it once we actually add ECDH support.

Buildbots are green, please let me know what you think.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @hynek<br>@exarkun set status to `new`|
|-|-|

  1. The docstring for `test_singleUseKeys` says what the test is doing rather than what the implementation is expected to do (though it does so without using the word _should_ for which I guess you get a few points :).
  1. The news fragment says "and makes servers choose the cipher".  I think this might be a little ambiguous wrt `CertificateOptions` usage for clients.  I think that `OP_CIPHER_SERVER_PREFERENCE` is a no-op on clients?  So maybe "and, for servers, uses server preference to choose the cipher."  That's a little awkward due to the repeated word "server" but maybe it's an improvement anyway; I'll leave it up to you.

Otherwise looks great, please address those issues and merge.  Thanks a lot!

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_6"></a>|@hynek set status to `closed`|
|-|-|

(In [41286]) Merge harden-tls-options-6801

Author: hynek
Reviewer: exarkun
Fixes: [#6801](https://github.com/twisted/twisted/issues/6801)

twisted.internet.ssl.CertificateOptions now disables TLS compression to avoid
CRIME attacks and, for servers, uses server preference to choose the cipher.
